### ISSUE TITLE
Multitask In Parallel on Browserstack

### ISSUE THREAD

**Issue by dtothefp** [webdriverio/webdriverio#614]
I realize this might be somewhat of a duplicate of @lucasfeliciano issue https://github.com/webdriverio/webdriverio/issues/301 but I would like to get more information on your comment @christian-bromann https://github.com/webdriverio/webdriverio/issues/301#issuecomment-56083306 about "What I usually do is configuring a browser matrix in my CI system and run each job in parallel using different environment variables".

I have to admit I'm a little disappointed with lack of documentation on BrowserStack integration with webdriverio on both the BrowserStack and webdriverio's website. I realize @christian-bromann works at SauceLabs, but it seems strange that BroswerStack demo's [WD](https://github.com/admc/wd) and not webriverio (even though I believe the WD maintainers are associated with SauceLabs as well). Regardless, I was impressed with webdriverio and the good documentation and github issue support by @christian-bromann.

Anyhow, I setup a repo that run's tests in 3 ways using `gulp` in combination with `gulp-mocha` and `webdriverio` https://github.com/dtothefp/browserstack-webdriverio-e2e/blob/master/gulp/tasks/selenium/index.js:
- `gulp selenium` uses `selenium-standalone` to download/start selenium server binary 
- `gulp selenium:tunnel` uses `browserstacktunnel-wrapper` allows testing of local IP's on browserstack (I like the [gulp-webdriver](https://github.com/webdriverio/gulp-webdriver) plugin but it seems to only support local tunnels with SauceLabs)
- `gulp selenium -e prod` runs tests on browserstack

I'm most interested in the last task which will be run on CI. It sounds like from your comments on https://github.com/webdriverio/webdriverio/issues/301 I may have to scrap using the webdriverio api in the test https://github.com/dtothefp/browserstack-webdriverio-e2e/blob/master/test/e2e/local-selenium-spec.js and use the CLI to spawn multiple jobs to run test on multiple browsers in parallel on BrowserStack?

The BrowserStack docs on parallel runs https://www.browserstack.com/automate/node#parallel-runs are quite poor so any info you could offer would be great.

**Comment by christian-bromann** [comment]
> I have to admit I'm a little disappointed with lack of documentation on BrowserStack integration with webdriverio

What would you suggest how we could improve that?

> I realize @christian-bromann works at SauceLabs

This is a private project of mine :wink: 

> Regardless, I was impressed with webdriverio and the good documentation and github issue support by @christian-bromann.

Thanks!

> use the CLI to spawn multiple jobs to run in parallel on BrowserStack?

I would recommend to use WebdriverIOs test runner `wdio`. It takes care about initialising multiple Selenium sessions on Browserstack or on your machine. The only thing you have to do when running it locally is to provide the Selenium server. You can also have different config files for each environment. It is currently the easiest way to run multiple tests in different browser at once. Just checkout the testrunner section on [our website](http://webdriver.io/guide.html).

Since you are using Gulp I would say you use a gulp plugin that executes the wdio runner or you help me to update [gulp-webdriver](https://github.com/webdriverio/gulp-webdriver) in order to have a nice plugin for that :wink:

**Comment by christian-bromann** [comment]
oh and what I meant in my comment was that I define my capabilities like this:

```js
desiredCapabilities: {
    browserName: process.env.BROWSER
}
```

and then execute my test suite multiple times with a different env variable. Somehow like travis does it with the environment matrix

**Comment by dtothefp** [comment]
Thanks for the reply and suggestions. I'm definitely happy to work on helping out on updating the gulp plugin. I haven't used the Travis matrix but it looks promising. Do you think something similar could be accomplished with node clusters with workers that individually spawn up child processes that execute the wdio runner with different browser config?

As for documentation what I had to dig for the most was getting the local tunnel working with BrowserStack which is what I think your gulp plugin does with Sauce. Also I believe their docs should mention webdriverio which is strange to be omitted when they mention WD and Nightwatch. I will definitely mention it to BrowserStack as we vet them for use at my company.

**Comment by christian-bromann** [comment]
> Do you think something similar could be accomplished with node clusters with workers that individually spawn up child processes that execute the wdio runner with different browser config?

Sure. There might be an NPM module that does something like this. Though last time I looked for it I haven't found anything

> I will definitely mention it to BrowserStack as we vet them for use at my company.

Cool thanks! I haven't contacted the guys from Browserstack yet. For me it's more important that the WebdriverIO docs are sufficient enough.

**Comment by dtothefp** [comment]
@christian-bromann I'm a little confused. It looks like you are starting many child process' based upon multiple capabilities listing in a wdio config file:

https://github.com/webdriverio/webdriverio/blob/master/lib/launcher.js#L23

Would this not run tests in parallel?

**Comment by dtothefp** [comment]
And that using the cli multiple capabilities can be configured

https://github.com/webdriverio/webdriverio/blob/master/examples/wdio.conf.js#L58

**Comment by christian-bromann** [comment]
@dtothefp yes if you define multiple capabilities in your wdio config tests get executed in parallel. That's why I recommended to use wdio to parallelise the test execution. The wdio test runner is something new that got introduced with `v3`.

**Comment by dtothefp** [comment]
@christian-bromann that's cool, I've been messing around with it and seems to work well except for using es6 in mocha which I usually do with the `require` flag in gulp mocha, where I require the following module

```js
require('babel/register')({
  experimental: true
});
```

But looks like this doesn't work with mocha for files that contain a `require()`. http://mochajs.org/#r-require-lt-module-name-gt

Looks like the workaround in gulp-mocha is to manually require the paths passed to it in an array https://github.com/sindresorhus/gulp-mocha/blob/master/index.js#L26

When I added this to your mocha framework code it works https://github.com/webdriverio/webdriverio/blob/master/lib/frameworks/mocha.js#L14. I'm basically using this because I don't think the `--harmony` flag supports arrow functions or es6 module syntax.

```js
if (opts.require && opts.require.length) {
  opts.require.forEach(require);
}

Regardless, I just discovered your docs on multiRemote http://webdriver.io/guide/usage/multiremote.html which looks like it runs tests in parallel as well. This looks great for our use case except seems like it adds extra assertions, and when running tests locally maybe I would only want to test one browser which would be problematic if there were more than on assertion for the multibrowser case. Also, I'm using the `chai-as-promised` library which I'm confused how to make work if the `.then` callback accepted more than one parameter for the multibrowser case. I saw that `chai-as-promised` looks highly customizable but I'm wondering if there is some way I could customize it to make one assertion per multibrowser test

**Comment by christian-bromann** [comment]
@dtothefp multiremote is supposed to enable integration tests with multiple browser so you can test your chat or web socket application.

If you want to use babel and ES6 just drop

```js
require('babel/register')({
  experimental: true
});
```

into your wdio config file and it should work.

**Comment by christian-bromann** [comment]
I guess this is fixed. Feel free to reopen if not.

**Comment by dtothefp** [comment]
hey @christian-bromann sorry in taking so long to follow up. I have a sample repo here https://github.com/dtothefp/webdriverio-parallel/tree/gh-pages that demonstrates using webdriverio in many different ways:
- with local selenium server `gulp selenium`
- tunnel into browserstack `gulp selenium:tunnel`
- testing remote IP on browserstack `gulp selenium -e prod`

and in parallel using a child process to spawn the wdio cli
- local selenium server `gulp selenium:parallel`
- local tunnel into browserstack `gulp selenium:tunnel:parallel`
- remote IP on browserstack `gulp selenium:parallel -e prod`

The tests are still a bit flaky other than on the local selenium server. I've also been experimenting in a more robust private repo with dynamic forms and have been progressively working the bugs out of the tests. 

Thanks for your help, I'm sure I'll have more questions/feedback as I go along

**Comment by saulshanabrook** [comment]
@christian-bromann So if I have a `wdio.config.js` with `capabilities` equal to:

```javascript
[
    {browserName: 'chrome', version: 'dev'},
    {browserName: 'chrome', version: 'beta'},
    {browserName: 'chrome', version: '44.0'}
  ]
```

And I run that through sauce labs then it will run all my tests in parallel?

**Comment by saulshanabrook** [comment]
I see that it does.

**Comment by christian-bromann** [comment]
@saulshanabrook @dtothefp :+1:

**Comment by gerbil** [comment]
And what about browserstack?
